### PR TITLE
[clang] Handle templated operators with reversed arguments

### LINKED ISSUE #53954: C++20 operator== ambiguous with derived-to-base conversion
```cpp
struct P {
#if 1
    template<class S>
    bool operator==(const S &) const;
#else
    template<class S>
    friend bool operator==(const P &, const S &);
#endif
};

struct A : public P {};
struct B : public P {};

bool check(A a, B b) {
    return a == b;
}
```

The `friend` function works just fine, but the member function dies with
```
<source>:15:14: error: use of overloaded operator '==' is ambiguous (with operand types 'A' and 'B')
    return a == b;
           ~ ^  ~
<source>:4:10: note: candidate function [with S = B]
    bool operator==(const S &) const;
         ^
<source>:4:10: note: candidate function (with reversed parameter order) [with S = A]
```

GCC and MSVC accept both. https://godbolt.org/z/9hj9554fn

@dwblaikie @zygoloid

### PR DESCRIPTION
**Pull request by usx95**
https://github.com/llvm/llvm-project/pull/68999 correctly computed conversion sequence for reversed args to a template operators. This was a breaking change as code, previously accepted in C++17, starts to break in C++20.

Example:
```cpp
struct P {};
template<class S> bool operator==(const P&, const S &);

struct A : public P {};
struct B : public P {};
bool check(A a, B b) { return a == b; }  // This is now ambiguous in C++20.
```

In order to minimise widespread breakages, as a clang extension, we had previously accepted such ambiguities with a warning (`-Wambiguous-reversed-operator`) for non-template operators. Due to the same reasons, we extend this relaxation for template operators.

Fixes https://github.com/llvm/llvm-project/issues/53954

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -37,6 +37,25 @@ These changes are ones which we think may surprise users when upgrading to
...
+
+    struct A : public P {};
+    struct B : public P {};
+
+    bool ambiguous(A a, B b) { return a == b; } // This equality is now ambiguous in C++20.
+
+    template<class S> bool operator!=(const P&, const S&);
+    bool fine(A a, B b) { return a == b; } // Ok. Found a matching operator!=.
+
+  To reduce widespread breakages, as an extension, clang would accept this with a
```
Comment (ilya-biryukov) [inline clang/docs/ReleaseNotes.rst:56]: NIT: s/clang/Clang to match the rest of the document.

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -37,6 +37,25 @@ These changes are ones which we think may surprise users when upgrading to
...
+  ``-Wambiguous-reversed-operator`` warning.
```
Comment (ilya-biryukov) [inline clang/docs/ReleaseNotes.rst:57]: NIT: it reads as if it is a new warning, I suggest saying "Clang accepts this code with an existing warning -W...."

File: clang/lib/Sema/SemaOverload.cpp [hunk 3]
```diff
@@ -7688,7 +7688,7 @@ bool Sema::CheckNonDependentConversions(
     QualType ParamType = ParamTypes[I + Offset];
     if (!ParamType->isDependentType()) {
       unsigned ConvIdx = PO == OverloadCandidateParamOrder::Reversed
-                             ? 0
+                             ? Args.size() - 1 - (ThisConversions + I)
```
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaOverload.cpp:7691]: Is this a separate bug? Do you have a code example that manifests it independently?
I suggest to move this to a separate commit with its own test if it's not directly connected.

File: clang/lib/Sema/SemaOverload.cpp [hunk 4]
```diff
@@ -10085,10 +10085,14 @@ getImplicitObjectParamType(ASTContext &Context, const FunctionDecl *F) {
   return M->getFunctionObjectParameterReferenceType();
 }
 
-static bool haveSameParameterTypes(ASTContext &Context, const FunctionDecl *F1,
+static bool allowAmbiguityWithSelf(ASTContext &Context, const FunctionDecl *F1,
```
Comment (ilya-biryukov) [inline clang/lib/Sema/SemaOverload.cpp:10088]: Could you elaborate why `declaresSameEntity` is not enough and we need extra checks for parameter types?
I am asking because if it's not enough for regular functions, maybe there is a similar case for function templates.
And the name is a bit misleading as `F1` and `F2` can clearly be two different functions, so it is unclear what `self`  refers to.

Comment (usx95) [reply clang/lib/Sema/SemaOverload.cpp:10088]: I added a todo to get rid of the parameter type checks. I do not see any test failures without it.
Since this could again be a potentially breaking change, I will try cleaning this up in a separate PR.

Changed the name to `allowAmbiguity`.

File: clang/lib/Sema/SemaOverload.cpp (hunk 3 above)
Comment (ilya-biryukov) [reply clang/lib/Sema/SemaOverload.cpp:7691]: Thanks for clearing this up for me offline. After going through an example I finally got what was wrong here.
For reversed operators `Args.size()` is always 2. When `ThisConversions == 0` (i.e. non-methods), we were writing conversions for *both* arguments into `ConvIdx == 0`, after this change we correctly loop through both.
When `ThisConversions == 1`, both versions work the same.

Indeed, the tests you added already check this behavior. Thanks! Resolving this.

**Review by ilya-biryukov — APPROVED** [review]
The change looks good, I believe the comments about potentially redundant checks and naming of the function could be addressed in a follow-up.

File: clang/lib/Sema/SemaOverload.cpp (hunk 4 above)
Comment (ilya-biryukov) [reply clang/lib/Sema/SemaOverload.cpp]: `allowAmbiguity` looks fine, thanks!

File: clang/lib/Sema/SemaOverload.cpp (hunk 3 above)
Comment (shafik) [reply clang/lib/Sema/SemaOverload.cpp:7691]: I think this warrants a comment as to why this computation is correct. It is certainly not obvious and if anyone ever needs to update this code it would be helpful.
